<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Indexed errors (purrr_error_indexed) — purrr_error_indexed • purrr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.9/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Indexed errors (purrr_error_indexed) — purrr_error_indexed"><meta name="description" content="The purrr_error_indexed class is thrown by map(), map2(), pmap(), and friends.
It wraps errors thrown during the processing on individual elements with information about the location of the error."><meta property="og:description" content="The purrr_error_indexed class is thrown by map(), map2(), pmap(), and friends.
It wraps errors thrown during the processing on individual elements with information about the location of the error."><meta property="og:image" content="https://purrr.tidyverse.org/logo.png"><meta name="robots" content="noindex"><script defer data-domain="purrr.tidyverse.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#container" class="visually-hidden-focusable">Skip to content</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-none" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">purrr</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.0.4.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/base.html">purrr &lt;-&gt; base R</a></li>
    <li><a class="dropdown-item" href="../articles/other-langs.html">Functional programming in other languages</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-news" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">News</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-news"><li><h6 class="dropdown-header" data-toc-skip>Releases</h6></li>
    <li><a class="external-link dropdown-item" href="https://www.tidyverse.org/blog/2022/12/purrr-1-0-0/">Version 1.0.0</a></li>
    <li><a class="external-link dropdown-item" href="https://www.tidyverse.org/articles/2019/02/purrr-0-3-0/">Version 0.3.0</a></li>
    <li><a class="external-link dropdown-item" href="https://www.tidyverse.org/articles/2017/08/purrr-0.2.3/">Version 0.2.3</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../news/index.html">Changelog</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/tidyverse/purrr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic" id="container">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Indexed errors (<code>purrr_error_indexed</code>)</h1>
      <small class="dont-index">Source: <a href="https://github.com/tidyverse/purrr/blob/main/R/map.R" class="external-link"><code>R/map.R</code></a></small>
      <div class="d-none name"><code>purrr_error_indexed.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The <code>purrr_error_indexed</code> class is thrown by <code><a href="map.html">map()</a></code>, <code><a href="map2.html">map2()</a></code>, <code><a href="pmap.html">pmap()</a></code>, and friends.
It wraps errors thrown during the processing on individual elements with information about the location of the error.</p>
    </div>


    <div class="section level2">
    <h2 id="structure">Structure<a class="anchor" aria-label="anchor" href="#structure"></a></h2>
    <p><code>purrr_error_indexed</code> has three important fields:</p><ul><li><p><code>location</code>: the location of the error as a single integer.</p></li>
<li><p><code>name</code>: the name of the location as a string. If the element was not named, <code>name</code> will be <code>NULL</code></p></li>
<li><p><code>parent</code>: the original error thrown by <code>.f</code>.</p></li>
</ul><p>Let's see this in action by capturing the generated condition from a very simple example:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/abort.html" class="external-link">abort</a></span><span class="op">(</span><span class="st">"This is an error"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">cnd</span> <span class="op">&lt;-</span> <span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/catch_cnd.html" class="external-link">catch_cnd</a></span><span class="op">(</span><span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">cnd</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "purrr_error_indexed" "rlang_error"         "error"</span></span>
<span><span class="co">#&gt; [4] "condition"</span></span>
<span></span>
<span><span class="va">cnd</span><span class="op">$</span><span class="va">location</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span></span>
<span><span class="va">cnd</span><span class="op">$</span><span class="va">name</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cnd</span><span class="op">$</span><span class="va">parent</span>, backtrace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;error/rlang_error&gt;</span></span>
<span><span class="co">#&gt; Error in `.f()`:</span></span>
<span><span class="co">#&gt; ! This is an error</span></span></code></pre><p></p></div>
<p>If the input vector is named, <code>name</code> will be non-<code>NULL</code>:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">cnd</span> <span class="op">&lt;-</span> <span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/catch_cnd.html" class="external-link">catch_cnd</a></span><span class="op">(</span><span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">1</span>, b <span class="op">=</span> <span class="fl">4</span>, c <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cnd</span><span class="op">$</span><span class="va">name</span></span>
<span><span class="co">#&gt; [1] "a"</span></span></code></pre><p></p></div>
    </div>
    <div class="section level2">
    <h2 id="handling-errors">Handling errors<a class="anchor" aria-label="anchor" href="#handling-errors"></a></h2>
    <p>(This section assumes that you're familiar with the basics of error handling in R, as described in <a href="https://adv-r.hadley.nz/conditions.html" class="external-link">Advanced R</a>.)</p>
<p>This error chaining is really useful when doing interactive data analysis, but it adds some extra complexity when handling errors with <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch()</a></code> or <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">withCallingHandlers()</a></code>.
Let's see what happens by adding a custom class to the error thrown by <code>f()</code>:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/abort.html" class="external-link">abort</a></span><span class="op">(</span><span class="st">"This is an error"</span>, class <span class="op">=</span> <span class="st">"my_error"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in `map()`:</span></span>
<span><span class="co">#&gt; i In index: 1.</span></span>
<span><span class="co">#&gt; Caused by error in `.f()`:</span></span>
<span><span class="co">#&gt; ! This is an error</span></span></code></pre><p></p></div>
<p>This doesn't change the visual display, but you might be surprised if you try to catch this error with <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch()</a></code> or <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">withCallingHandlers()</a></code>:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span>,</span>
<span>  my_error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">err</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># use NULL value if error</span></span>
<span>    <span class="cn">NULL</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in `map()`:</span></span>
<span><span class="co">#&gt; i In index: 1.</span></span>
<span><span class="co">#&gt; Caused by error in `.f()`:</span></span>
<span><span class="co">#&gt; ! This is an error</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">withCallingHandlers</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span>,</span>
<span>  my_error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">err</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># throw a more informative error</span></span>
<span>    <span class="fu">abort</span><span class="op">(</span><span class="st">"Wrapped error"</span>, parent <span class="op">=</span> <span class="va">err</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in `map()`:</span></span>
<span><span class="co">#&gt; i In index: 1.</span></span>
<span><span class="co">#&gt; Caused by error in `.f()`:</span></span>
<span><span class="co">#&gt; ! This is an error</span></span></code></pre><p></p></div>
<p>That's because, as described above, the error that <code><a href="map.html">map()</a></code> throws will always have class <code>purrr_error_indexed</code>:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span>,</span>
<span>  purrr_error_indexed <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">err</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Hello! I am now called :)"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Hello! I am now called :)"</span></span></code></pre><p></p></div>
<p>In order to handle the error thrown by <code>f()</code>, you'll need to use <code><a href="https://rlang.r-lib.org/reference/cnd_inherits.html" class="external-link">rlang::cnd_inherits()</a></code> on the parent error:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span>,</span>
<span>  purrr_error_indexed <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">err</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/cnd_inherits.html" class="external-link">cnd_inherits</a></span><span class="op">(</span><span class="va">err</span>, <span class="st">"my_error"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="cn">NULL</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>      <span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/cnd_signal.html" class="external-link">cnd_signal</a></span><span class="op">(</span><span class="va">err</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">withCallingHandlers</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span>,</span>
<span>  purrr_error_indexed <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">err</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/cnd_inherits.html" class="external-link">cnd_inherits</a></span><span class="op">(</span><span class="va">err</span>, <span class="st">"my_error"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu">abort</span><span class="op">(</span><span class="st">"Wrapped error"</span>, parent <span class="op">=</span> <span class="va">err</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Error:</span></span>
<span><span class="co">#&gt; ! Wrapped error</span></span>
<span><span class="co">#&gt; Caused by error in `map()`:</span></span>
<span><span class="co">#&gt; i In index: 1.</span></span>
<span><span class="co">#&gt; Caused by error in `.f()`:</span></span>
<span><span class="co">#&gt; ! This is an error</span></span></code></pre><p></p></div>
<p>(The <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch()</a></code> approach is suboptimal because we're no longer just handling errors, but also rethrowing them.
The rethrown errors won't work correctly with (e.g.) <code><a href="https://rdrr.io/r/utils/recover.html" class="external-link">recover()</a></code> and <code><a href="https://rdrr.io/r/base/traceback.html" class="external-link">traceback()</a></code>, but we don't currently have a better approach.
In the future we expect to <a href="https://github.com/r-lib/rlang/issues/1534" class="external-link">enhance <code>try_fetch()</code></a> to make this easier to do 100% correctly).</p>
<p>Finally, if you just want to get rid of purrr's wrapper error, you can resignal the parent error:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">withCallingHandlers</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>, <span class="va">f</span><span class="op">)</span>,</span>
<span>  purrr_error_indexed <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">err</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/cnd_signal.html" class="external-link">cnd_signal</a></span><span class="op">(</span><span class="va">err</span><span class="op">$</span><span class="va">parent</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in `.f()`:</span></span>
<span><span class="co">#&gt; ! This is an error</span></span></code></pre><p></p></div>
<p>Because we are resignalling an error, it's important to use <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">withCallingHandlers()</a></code> and not <code><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch()</a></code> in order to preserve the full backtrace context.
That way <code><a href="https://rdrr.io/r/utils/recover.html" class="external-link">recover()</a></code>, <code><a href="https://rdrr.io/r/base/traceback.html" class="external-link">traceback()</a></code>, and related tools will continue to work correctly.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


   </div>
  <footer><div class="container">
  <div class="pkgdown-footer-left">
  <p>Developed by <a href="https://hadley.nz" class="external-link">Hadley Wickham</a>, <a href="https://github.com/lionel-" class="external-link">Lionel Henry</a>, <a href="https://www.posit.co" class="external-link"><img src="https://www.tidyverse.org/posit-logo.svg" alt="Posit" height="16" width="62" style="margin-bottom: 3px;"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

  </div></footer></body></html>

